### PR TITLE
[#15] [Android] [UI] As a user, I can see a Splash screen when app starts

### LINKED ISSUE #15: [Android] [UI] As a user, I can see the Splash screen
# Why

To build the application identity and keeps users occupied while your app loads in the background, we will show the `Splash` screen when the app is starting.

# Acceptance Criteria
- Show a background image and the `Nimble` as the following design

Image

[Figma](https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=31-0&t=lF1bJOT4aAdBAohd-0)

### PR DESCRIPTION
**Pull request by AVI5HEK**
- Close #15 

## What happened 👀

To build the application identity and keeps users occupied while your app loads in the background, we will show the Splash screen when the app is starting.

## Insight 📝

- Created content for the Splash screen.
- Made the background, overlay, and logo separate Composables to reuse them in the Login screen.
- Added app logo and name.
- Renamed theme.

## Proof Of Work 📹

https://user-images.githubusercontent.com/8093908/233991243-28b58d7d-6a49-4b56-8648-de7055737608.mp4

### CODE REVIEW

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/MainActivity.kt [hunk 1]
```diff
@@ -9,17 +9,20 @@ import androidx.compose.runtime.Composable
...
         super.onCreate(savedInstanceState)
         setContent {
-            MyApplicationTheme {
+            ApplicationTheme {
                 Surface(
                     modifier = Modifier.fillMaxSize(),
                     color = MaterialTheme.colors.background
                 ) {
-                    GreetingView(Greeting().greet())
+                    //GreetingView(Greeting().greet())
```
Comment (suho) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/MainActivity.kt]: We can remove this view

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import androidx.compose.foundation.Image
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.res.painterResource
+import co.nimblehq.avishek.phong.kmmic.android.R
+
+@Composable
+fun Background(modifier: Modifier) {
```
Comment (suho) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: Modifier should be an optional parameter
```suggestion
fun Background(modifier: Modifier = Modifier) {
```

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Brush
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.platform.LocalConfiguration
+
+@Composable
+fun Overlay(
+    modifier: Modifier,
+    colors: List<Color>
+) {
```
Comment (suho) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt]: Same here. The modifier should be an optional parameter:

```suggestion
fun Overlay(
    colors: List<Color>,
	modifier: Modifier = Modifier,
) {
```

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    SplashContent(shouldShowLogo = shouldShowLogo)
+}
+
+@Composable
+fun SplashContent(
+    modifier: Modifier = Modifier,
+    shouldShowLogo: Boolean,
+) {
```
Comment (suho) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt]:
```suggestion
fun SplashContent(
    shouldShowLogo: Boolean,
	modifier: Modifier = Modifier,
) {
```

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/MainActivity.kt (hunk 1 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/MainActivity.kt]: Resolved in acd211abce503054711d0c24171bd83cc8117768

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt (hunk 2 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: Added in acd211abce503054711d0c24171bd83cc8117768

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt (hunk 3 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt]: Fixed in acd211abce503054711d0c24171bd83cc8117768

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt (hunk 4 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt]: Fixed in acd211abce503054711d0c24171bd83cc8117768

File: androidApp/src/main/res/mipmap-anydpi-v26/ic_launcher_round.xml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
```
[the remark this answers is not in the document]
Comment (AVI5HEK) [reply androidApp/src/main/res/mipmap-anydpi-v26/ic_launcher_round.xml:2]: This is a false positive. We can safely ignore it.

File: androidApp/src/main/res/drawable/logo_white.xml [hunk 6]
```diff
@@ -0,0 +1,94 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="202dp"
```
[the remark this answers is not in the document]
Comment (AVI5HEK) [reply androidApp/src/main/res/drawable/logo_white.xml:2]: This is imported as-is from Figma.

File: androidApp/src/main/res/drawable/logo_white.xml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    android:height="48dp"
+    android:viewportWidth="202"
+    android:viewportHeight="48">
+  <group>
+    <clip-path
```
[the remark this answers is not in the document]
Comment (AVI5HEK) [reply androidApp/src/main/res/drawable/logo_white.xml:7]: This is imported from Figma. Let's just hope it'll work.

File: androidApp/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
     <application
         android:allowBackup="false"
```
[the remark this answers is not in the document]
Comment (AVI5HEK) [reply androidApp/src/main/AndroidManifest.xml:5]: This is autogenerated by Android Studio during the project bootstrapping and out of scope of this PR.

**Review by suho — APPROVED** [review]

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+import androidx.compose.foundation.Image
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.res.painterResource
+import co.nimblehq.avishek.phong.kmmic.android.R
+
+@Composable
+fun Background(modifier: Modifier = Modifier) {
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: Please avoid separating Composables into different files unless they are `commonly used by other Composables` or `has multiple UI states`. [ref](https://nimblehq.co/compass/development/android/user-interface/jetpack-compose/#separating-compose-functions-into-different-files)

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Logo.kt [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+import androidx.compose.foundation.Image
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.res.painterResource
+import co.nimblehq.avishek.phong.kmmic.android.R
+
+@Composable
+fun Logo(modifier: Modifier = Modifier) {
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Logo.kt]: Same here.

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Box
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Brush
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.platform.LocalConfiguration
+
+@Composable
+fun Overlay(
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt]: This too.

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt (hunk 9 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: @kaungkhantsoe I plan to reuse them on the Login screen, which is separate from the Splash screen.

Comment (kaungkhantsoe) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: @AVI5HEK Still, it is not worth separating this view into a separate file. You can check the parameters ☝️. Separating into different Composable means that you want to do `State hoisting` to reduce the recomposing times or it is reused by other composable functions and it has more parameters than `Modifier`. Here we are just separating simple views like `Image` into different files which is not a best practice. Same for `Logo` and `Overlay`. Imagine separating `<Image/>` into different `xml` file and latter `<include/>` into different layouts. It is weird and unnecessary right?

Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: @kaungkhantsoe Separating the Splash and Login screens also didn't work. So I'm going to combine them together in one screen. The names will be updated along with some other refactoring.

Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Background.kt]: Updated in 09abeb08dec08b3959cd0e8f2837267c041bde8b

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Logo.kt (hunk 10 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Logo.kt]: Updated in 09abeb08dec08b3959cd0e8f2837267c041bde8b

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt (hunk 11 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/common/Overlay.kt]: Updated in 09abeb08dec08b3959cd0e8f2837267c041bde8b

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+fun SplashContent(
+    shouldShowLogo: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    Box(
+        contentAlignment = Alignment.Center,
+        modifier = modifier.fillMaxSize()
+    ) {
+        Image(
+            painterResource(id = R.drawable.bg_splash),
```
Comment (kaungkhantsoe) [inline androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt]:
```suggestion
            painter = painterResource(id = R.drawable.bg_splash),
```

**Review by kaungkhantsoe — COMMENTED** [review]
One last concern. The rest LGTM

File: androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt (hunk 12 above)
Comment (AVI5HEK) [reply androidApp/src/main/java/co/nimblehq/avishek/phong/kmmic/android/ui/screen/login/SplashScreen.kt]: @kaungkhantsoe Resolved in https://github.com/nimblehq/avishek-phong-kmm-ic/pull/80/commits/548205d70b6e29a6397e32576e315355c5e9ed21.

**Review by kaungkhantsoe — APPROVED** [review]

**Pull request merged**